File: src/modifiers/animation.ts
import { createModifier } from './createModifier';

export type AnimationCurve =
  | 'default'
  | 'easeIn'
  | 'easeOut'
  | 'easeInOut'
  | 'linear'
  | 'spring'
  | 'none';

export interface AnimationParams {
  /// Curve preset. Defaults to `'default'`. Use `'none'` to disable
  /// animation (matches SwiftUI's `nil` animation argument).
  type?: AnimationCurve;
  /// Override duration in seconds. Honored by curves that accept one
  /// (`easeIn`, `easeOut`, `easeInOut`, `linear`). Ignored by `spring`.
  duration?: number;
  /// Stable hashable key. SwiftUI animates implicit changes whenever this
  /// value differs from the previous render. Pass the same value across
  /// renders to keep the animation idle; bump it (or pass the changing
  /// value itself) when an animation should fire.
  value?: string | number | boolean;
}

/// SwiftUI `.animation(_:value:)`. Wraps any view; subsequent prop or
/// child changes that happen on the same commit as a `value` change
/// animate with the chosen curve.
///
/// ```tsx
/// <ProgressView
///   value={pct}
///   modifiers={[animation({ type: 'easeIn', value: pct })]}
/// />
/// ```
export function animation(params: AnimationParams = {}) {
  return createModifier('animation', params);
}

File: apple/Sources/ReactNativeWatchOS/Modifiers/Animation.swift
import SwiftUI

/// SwiftUI `.animation(_:value:)`. The animation curve replays whenever
/// `value` changes between renders — JS passes a stable hashable key
/// (number / string / bool) and bumps it when an animation should fire.
///
/// `type` selects the curve; `duration` overrides the default duration
/// for curves that accept one. `type: "none"` passes `nil` to disable
/// animation for this segment (matches SwiftUI's nil-animation semantics).
enum RNWAnimationModifier {
    @MainActor
    static func register(into r: RNWModifierRegistry) {
        r.register("animation") { view, params, _ in
            let type = params.string("type") ?? "default"
            let duration = params.double("duration")
            let valueKey = AnimationValueKey(raw: params["value"])
            let animation = makeAnimation(type: type, duration: duration)
            return AnyView(view.animation(animation, value: valueKey))
        }
    }

    private static func makeAnimation(
        type: String, duration: Double?
    ) -> Animation? {
        switch type {
        case "none":
            return nil
        case "easeIn":
            return duration.map { .easeIn(duration: $0) } ?? .easeIn
        case "easeOut":
            return duration.map { .easeOut(duration: $0) } ?? .easeOut
        case "easeInOut":
            return duration.map { .easeInOut(duration: $0) } ?? .easeInOut
        case "linear":
            return duration.map { .linear(duration: $0) } ?? .linear
        case "spring":
            return .spring()
        default:
            return .default
        }
    }
}

/// Stable Hashable wrapping any JS-side `value`. SwiftUI compares this
/// across renders to decide whether to fire the animation. We coerce
/// the raw bridged value into a single string — exact bool/int parity
/// doesn't matter, only that the same input always hashes the same way.
private struct AnimationValueKey: Hashable {
    let key: String

    init(raw: Any?) {
        switch raw {
        case .none:
            key = ""
        case let s as String:
            key = "s:" + s
        case let n as NSNumber:
            key = "n:" + n.stringValue
        case let other?:
            key = "x:" + String(describing: other)
        }
    }
}
